radix-engine/src/model/access_controller/state_machine.rs

```diff
@@ -0,0 +1,421 @@
+        api: &mut Y,
+        input: AccessControllerTimedConfirmRecoveryStateMachineInput,
+    ) -> Result<Self::Output, RuntimeError>
+    where
+        Y: SystemApi + EngineApi<RuntimeError> + InvokableModel<RuntimeError>,
+    {
+        // Timed confirm recovery can only be performed by the recovery role (this is checked
+        // through access rules on the invocation itself) and can be performed in recovery mode
+        // regardless of whether primary is locked or unlocked.
+        match self.state {
+            (
+                _,
+                _,
+                RecoveryOperationState::Recovery(RecoveryRecoveryState::Timed {
+                    ref proposal,
+                    ref timed_recovery_allowed_after,
+                }),
+            ) => {
+                let proposal = proposal.clone();
+
+                // Ensure that the caller has passed in the expected proposal
+                validate_recovery_proposal(&proposal, &input.proposal_to_confirm)?;
+
+                let recovery_time_has_elapsed = Runtime::sys_compare_against_current_time(
+                    api,
+                    timed_recovery_allowed_after.clone(),
+                    TimePrecision::Minute,
+                    TimeComparisonOperator::Gte,
+                )?;
+
+                // If the timed recovery delay has elapsed, then we transition into normal
+                // operations mode with primary unlocked and return the ruleset that was found.
+                if !recovery_time_has_elapsed {
+                    access_controller_runtime_error!(TimedRecoveryDelayHasNotElapsed)
+                } else {
+                    self.state = Default::default();
+
+                    Ok(proposal)
+                }
+            }
+            _ => access_controller_runtime_error!(NoTimedRecoveriesFound),
+        }
+    }
+}
+
+pub(super) struct AccessControllerCancelRecoveryAttemptStateMachineInput {
+    pub proposer: Proposer,
+    pub proposal_to_cancel: RecoveryProposal,
+}
+
+impl TransitionMut<AccessControllerCancelRecoveryAttemptStateMachineInput>
+    for AccessControllerSubstate
+{
+    type Output = ();
+
+    fn transition_mut<Y>(
+        &mut self,
+        _api: &mut Y,
+        input: AccessControllerCancelRecoveryAttemptStateMachineInput,
+    ) -> Result<Self::Output, RuntimeError>
+    where
+        Y: SystemApi + EngineApi<RuntimeError> + InvokableModel<RuntimeError>,
+    {
+        // A recovery attempt can only be canceled when we're in recovery mode regardless of whether
+        // primary is locked or unlocked
+        match self.state {
+            (_, PrimaryOperationState::Recovery(ref proposal), _)
+                if input.proposer == Proposer::Primary =>
+            {
+                let proposal = proposal.clone();
+
+                // Ensure that the caller has passed in the expected proposal
+                validate_recovery_proposal(&proposal, &input.proposal_to_cancel)?;
```

Comment:
Is this necessary for cancelling?

Reply:
I guess it isn't. CAP-14 primarily mentions passing the proposal again for quick confirm recovery but not for cancellation of a recovery proposal, so this might not be necessary.